src/redprl/refiner_types.fun

```diff
@@ -2040,24 +2057,42 @@ struct
         val x = alpha 0
         val truncatedHx = truncatedH @> (x, AJ.TRUE holeTy)
         val (motiveGoal, motiveHole) = makeTerm truncatedHx O.EXP
-        val motiveWfGoal = makeType truncatedHx (motiveHole, K.top)
+        val motiveWfGoal = View.makeAsMem truncatedHx (motiveHole, ty)
 
         val motiven = substVar (holeN, x) motiveHole
         val motivem = substVar (holeM, x) motiveHole
 
-        fun replace jdg = 
-          case jdg of 
-             AJ.TRUE _ => AJ.TRUE motiven
-           | _ => jdg
-
-        val (H', ajdg') = Selector.map sel replace (H, ajdg)
-        val (rewrittenGoal, rewrittenHole) = makeGoal @@ H' >> ajdg'
+        val replace =
```

Comment:
@favonia Do you think that there is any way that this kind of code can be factored out? If it sounds like it is too difficult or won't help, don't worry about it.

Reply:
I tried, but then decided to write out the actual program before refactoring. Any ideas?

Reply:
The difficulty is that the semantics is not independent of rewriting. For example, `SYNTH` should be dealt with in general map/project functions, but it is not mentioned here because it is unclear what the criterion of equality should be for a term with an unknown type.